File: elixir/lfpwe/ch01/myset.exs
defmodule Example do
  defmodule MySet do
    defstruct items: []

    def push(set = %{items: items}, item) do
      if Enum.member?(items, item) do
        set
      else
        %{set | items: items ++ [item]}
      end
    end
  end

  def run do
    set = %MySet{}
    set = MySet.push(set, "apple")

    new_set = %MySet{}
    new_set = MySet.push(new_set, "pie")

    IO.inspect MySet.push(set, "apple")

    IO.inspect MySet.push(new_set, "apple")

    ["dogs", "cats", "flowers"]
    |> Enum.map(&String.upcase/1)
    |> IO.inspect
  end
end

Example.run

File: elixir/lfpwe/ch01/myset.rb
class MySet
  attr_reader :items

  def initialize()
    @items = []
  end

  def push(item)
    items.push(item) unless items.include?(item)
  end
end

set = MySet.new
set.push("apple")

new_set = MySet.new
new_set.push("pie")

set.push("apple")
print set.items
new_set.push("apple")
print new_set.items
